{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DiffPaSS â€“ Differentiable Pairing using Soft Scores\n",
    "\n",
    "> Differentiable Pairing using Soft Scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"https://raw.githubusercontent.com/Bitbol-Lab/DiffPaSS/main/media/DiffPaSS_logo.svg\" width=\"640\" height=\"320\" alt=\"DiffPaSS logo\" />\n",
    "\n",
    "## Overview\n",
    "\n",
    "DiffPaSS is a family of high-performance and scalable PyTorch modules for finding optimal one-to-one pairings between two collections of biological sequences, and for performing general graph alignment.\n",
    "\n",
    "### Pairing multiple-sequence alignments (MSAs)\n",
    "\n",
    "A typical example of the problem DiffPaSS is designed to solve is the following: given two multiple sequence alignments (MSAs) A and B, containing interacting biological sequences, find the optimal one-to-one pairing between the sequences in A and B.\n",
    "\n",
    "<figure>\n",
    "  <img src=\"https://raw.githubusercontent.com/Bitbol-Lab/DiffPaSS/main/media/MSA_pairing_problem.svg\" width=\"640\" height=\"201.6\" alt=\"MSA pairing problem\" />\n",
    "  <figcaption>Pairing problem for two multiple sequence alignments, where pairings are restricted to be within the same species</figcaption>\n",
    "</figure>\n",
    "\n",
    "To find an optimal pairing, we can maximize the average mutual information between columns of the two paired MSAs (`InformationPairing`), or we can maximize the similarity between distance-based (`MirrortreePairing`) or orthology (`BestHitsPairing`) networks constructed from the two MSAs.\n",
    "\n",
    "### Graph alignment and pairing unaligned sequence collections\n",
    "\n",
    "DiffPaSS can be used for general graph alignment problems (`GraphAlignment`), where the goal is to find the one-to-one pairing between the nodes of two weighted graphs that maximizes the similarity between the two graphs. The user can specify the (dis-)similarity measure to be optimized, as an arbitrary differentiable function of the adjacency matrices of the two graphs.\n",
    "\n",
    "Using this capability, DiffPaSS can be used for finding the optimal one-to-one pairing between two unaligned collections of sequences, if weighted graphs are built in advance from the two collections (for example, using the pairwise Levenshtein distance). This is useful when alignments are not available or reliable.\n",
    "\n",
    "### Can I pair two collections with a different number of sequences?\n",
    "\n",
    "DiffPaSS optimizes and returns permutation matrices. Hence, its inputs are required to have the same number of sequences. However, DiffPaSS can be used to pair two collections (e.g. MSAs) containing a different number of sequences, by padding the smaller collection with dummy sequences. For multiple sequence alignments, a simple choice is to add dummy sequences consisting entirely of gap symbols. For general graphs, dummy nodes, connected to the other nodes with arbitrary edge weights, can be added to the smaller graph.\n",
    "\n",
    "### How DiffPaSS works: soft scores, differentiable optimization, bootstrap\n",
    "\n",
    "Check [our paper](https://openreview.net/forum?id=n5hO5seROB) for details of the DiffPaSS and DiffPaSS-IPA algorithms. Briefly, the main ingredients are as follows:\n",
    "\n",
    "  1. Using \"soft\" scores that differentiably extend information-theoretic scores between two paired multiple sequence alignments (MSAs), or scores based on sequence similarity or graph similarity measures.\n",
    "\n",
    "  2. The (truncated) Sinkhorn operator for smoothly parametrizing \"soft permutations\", and the matching operator for parametrizing real permutations [[Mena et al, 2018]](https://openreview.net/forum?id=Byt3oJ-0W).\n",
    "\n",
    "  3. A novel and efficient bootstrap technique, motivated by mathematical results and heuristic insights into this smooth optimization process. See the animation below for an illustration.\n",
    "\n",
    "  4. A notion of \"robust pairs\" that can be used to identify pairs that are consistently found throughout a DiffPaSS bootstrap. These pairs can be used as ground truths in another DiffPaSS run, giving rise to the DiffPaSS-Iterative Pairing Algorithm (DiffPaSS-IPA).\n",
    "  \n",
    "<figure>\n",
    "    <video src=\"https://github.com/Bitbol-Lab/DiffPaSS/assets/46537483/e411fe8c-2fed-4723-a25c-ff69a1abccec\" width=\"640\" height=\"360\" controls></video>\n",
    "    <figcaption>The DiffPaSS bootstrap technique and robust pairs</figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation\n",
    "\n",
    "### From PyPI\n",
    "\n",
    "DiffPaSS requires Python 3.7 or later. It is available on PyPI and can be installed with ``pip``:\n",
    "\n",
    "```sh\n",
    "python -m pip install diffpass\n",
    "```\n",
    "\n",
    "### From source\n",
    "\n",
    "Clone this repository on your local machine by running\n",
    "\n",
    "```sh\n",
    "git clone git@github.com:Bitbol-Lab/DiffPaSS.git\n",
    "```\n",
    "\n",
    "and move inside the root folder. We recommend creating and activating a dedicated conda or virtualenv Python virtual environment. Then, make an editable install of the package:\n",
    "\n",
    "``` sh\n",
    "python -m pip install -e .\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quickstart\n",
    "\n",
    "### Input data preprocessing (MSA pairing)\n",
    "\n",
    "First, parse your multiple sequence alignments (MSAs) in FASTA format into a list of tuples ``(header, sequence)`` using `read_msa`.\n",
    "\n",
    "```python\n",
    "from diffpass.msa_parsing import read_msa\n",
    "\n",
    "# Parse the MSAs into lists of tuples (header, sequence)\n",
    "msa_A = read_msa(\"path/to/msa_A.fasta\")\n",
    "msa_B = read_msa(\"path/to/msa_B.fasta\")\n",
    "```\n",
    "\n",
    "We assume that the MSAs contain species information in the headers, which will be used to restrict the pairings to be within the same species (more generally, \"groups\"). We need a simple function to extract the species information from the headers. For instance, if the headers are in the format `>sequence_id|species_name|...`, we can use:\n",
    "\n",
    "```python\n",
    "def species_name_func(header):\n",
    "    return header.split(\"|\")[1]\n",
    "```\n",
    "\n",
    "This function will be used to group the sequences by species:\n",
    "\n",
    "```python\n",
    "from diffpass.data_utils import create_groupwise_seq_records\n",
    "\n",
    "msa_A_by_sp = create_groupwise_seq_records(msa_A, species_name_func)\n",
    "msa_B_by_sp = create_groupwise_seq_records(msa_B, species_name_func)\n",
    "```\n",
    "\n",
    "If one of the MSAs contains sequences from species not present in the other MSA, we can remove these species from both MSAs:\n",
    "\n",
    "```python\n",
    "from diffpass.data_utils import remove_groups_not_in_both\n",
    "\n",
    "msa_A_by_sp, msa_B_by_sp = remove_groups_not_in_both(\n",
    "    msa_A_by_sp, msa_B_by_sp\n",
    ")\n",
    "```\n",
    "\n",
    "If there are species with different numbers of sequences in the two MSAs, we can add dummy sequences to the smaller species to make the number of sequences equal. For example, we can add dummy sequences consisting entirely of gap symbols:\n",
    "\n",
    "```python\n",
    "from diffpass.data_utils import pad_msas_with_dummy_sequences\n",
    "\n",
    "msa_A_by_sp_pad, msa_B_by_sp_pad = pad_msas_with_dummy_sequences(\n",
    "    msa_A_by_sp, msa_B_by_sp\n",
    ")\n",
    "\n",
    "species = list(msa_A_by_sp_pad.keys())\n",
    "species_sizes = list(map(len, msa_A_by_sp_pad.values()))\n",
    "```\n",
    "\n",
    "Next, one-hot encode the MSAs using the `one_hot_encode_msa` function.\n",
    "\n",
    "```python\n",
    "from diffpass.data_utils import one_hot_encode_msa\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "\n",
    "# Unpack the padded MSAs into a list of records\n",
    "msa_A_for_pairing = [\n",
    "    rec for recs_this_sp in msa_A_by_sp_pad.values() for rec in recs_this_sp\n",
    "]\n",
    "msa_B_for_pairing = [\n",
    "    rec for recs_this_sp in msa_B_by_sp_pad.values() for rec in recs_this_sp\n",
    "]\n",
    "\n",
    "# One-hot encode the MSAs and load them to a device\n",
    "msa_A_oh = one_hot_encode_msa(msa_A_for_pairing, device=device)\n",
    "msa_B_oh = one_hot_encode_msa(msa_B_for_pairing, device=device)\n",
    "```\n",
    "\n",
    "### Pairing optimization\n",
    "\n",
    "Finally, we can instantiate a class from `diffpass.train` to find an optimal pairing between `x` and `y`. Here, `x` and `y` are MSAs, so we can look for a pairing that optimizes the mutual information between `x` and `y`. For this, we use `InformationPairing` and the DiffPaSS bootstrapped optimization algorithm. See the tutorials below for other examples, including for graph alignment when `x` and `y` are weighted adjacency matrices.\n",
    "\n",
    "```python\n",
    "from diffpass.train import InformationPairing\n",
    "\n",
    "information_pairing = InformationPairing(group_sizes=species_sizes).to(device)\n",
    "bootstrap_results = information_pairing.fit_bootstrap(x, y)\n",
    "```\n",
    "\n",
    "The results are stored in a `DiffPaSSResults` container. The lists of (hard) losses and permutations found during the optimization can be accessed as attributes of the container:\n",
    "```python\n",
    "print(f\"Final hard loss: {bootstrap_results.hard_losses[-1].item()}\")\n",
    "print(f\"Final hard permutations (one permutation per species): {bootstrap_results.hard_perms[-1][-1].item()}\")\n",
    "```\n",
    "\n",
    "For more details and examples, including the DiffPaSS-IPA variant, see the tutorials."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tutorials\n",
    "\n",
    "- [`mutual_information_msa_pairing.ipynb`](https://github.com/Bitbol-Lab/DiffPaSS/blob/main/nbs/tutorials/mutual_information_msa_pairing.ipynb): paired MSA optimization using mutual information in the case of well-known prokaryotic datasets, for which ground truth pairings are given by genome proximity.\n",
    "- [`graph_alignment.ipynb`](https://github.com/Bitbol-Lab/DiffPaSS/blob/main/nbs/tutorials/graph_alignment.ipynb): general graph alignment using `diffpass.train.GraphAlignment`, with an example of aligning two weighted adjacency matrices."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Documentation\n",
    "\n",
    "The full documentation is available at [https://Bitbol-Lab.github.io/DiffPaSS/](https://bitbol-lab.github.io/DiffPaSS/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Citation\n",
    "\n",
    "To cite this work, please refer to the following publication:\n",
    "\n",
    "```bibtex\n",
    "@inproceedings{\n",
    "  lupo2024diffpass,\n",
    "  title={{DiffPaSS} -- {D}ifferentiable and scalable pairing of biological sequences using soft scores},\n",
    "  author={Umberto Lupo and Damiano Sgarbossa and Martina Milighetti and Anne-Florence Bitbol},\n",
    "  booktitle={ICLR 2024 Workshop on Generative and Experimental Perspectives for Biomolecular Design},\n",
    "  year={2024},\n",
    "  url={https://openreview.net/forum?id=n5hO5seROB}\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## nbdev\n",
    "\n",
    "Project developed using [nbdev](https://nbdev.fast.ai/)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
